File: common/src/main/java/de/hype/bbsentials/shared/objects/ChestLobbyData.java
package de.hype.bbsentials.shared.objects;

import de.hype.bbsentials.shared.constants.StatusConstants;

import java.awt.*;
import java.sql.SQLException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class ChestLobbyData {
    public String contactMan;
    public List<ChChestData> chests = new ArrayList<>();
    public String bbcommand;
    public String extraMessage;
    public transient Color color;
    public int lobbyId;
    public String serverId;
    protected String status;
    protected List<String> playersStillIn = new ArrayList<>();
    protected Instant closingTime;

    public ChestLobbyData(Collection<ChChestData> chest, String serverId, String bbcommand, String extraMessage, Object status) {
        chests.addAll(chest);
        this.serverId = serverId;
        this.contactMan = chests.get(0).finder;
        this.bbcommand = bbcommand;
        this.extraMessage = extraMessage;
        setStatusNoOverride(status);
    }

    public String getStatus() {
        return status;
    }

    /**
     * @param statusBase String or StatusConstants as buttonStyle.
     * @throws IllegalArgumentException if Object is not a {@link String} or {@link StatusConstants}
     */
    public void setStatus(Object statusBase) throws SQLException {
        setStatusNoOverride(statusBase);
    }

    public void setStatus(StatusConstants statusBase) throws SQLException {
        setStatusNoOverride(statusBase);
    }

    public void setStatusNoOverride(Object statusBase) {
        if (statusBase instanceof StatusConstants statusConstants) {
            this.status = statusConstants.getDisplayName();
            color = statusConstants.getColor();
        }
        else if (statusBase instanceof String) {
            this.status = (String) statusBase;
        }
        else {
            throw new IllegalArgumentException("Invalid input buttonStyle. Expected String or StatusConstants.");
        }
    }

    public void addChest(ChChestData chest) {
        chests.add(chest);
    }

    public void transferToUser(String newContactMan) {
        contactMan = newContactMan;
    }

    public void setLobbyMetaData(List<String> playersStillIn, Instant closingTime) throws SQLException {
        if (playersStillIn != null) {
            this.playersStillIn = playersStillIn;
        }
        if (closingTime != null) {
            this.closingTime = closingTime;
            onLobbyUpdate();
        }
    }

    public void onLobbyUpdate() {
        // need to be overridden
    }

    public List<String> getPlayersStillIn() {
        return playersStillIn;
    }

    public Instant getClosingTime() {
        return closingTime;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof ChestLobbyData)) return false;
        return ((ChestLobbyData) obj).lobbyId == lobbyId;
    }

    protected void updateLobby(ChestLobbyData lobby) {
        bbcommand = lobby.bbcommand;
        extraMessage = lobby.extraMessage;
        status = lobby.getStatus();
        contactMan = lobby.contactMan;
        chests = new ArrayList<>(lobby.chests);
        try {
            setLobbyMetaData(lobby.getPlayersStillIn(), lobby.getClosingTime());
        } catch (SQLException ignored) {
        }
    }
}

File: common/src/main/kotlin/de/hype/bingonet/shared/objects/ChestLobbyData.kt
package de.hype.bingonet.shared.objects

import de.hype.bingonet.shared.constants.StatusConstants
import java.awt.Color
import java.sql.SQLException
import java.time.Instant
import java.time.temporal.ChronoUnit

open class ChestLobbyData(
    chests: MutableList<ChChestData>,
    serverId: String,
    bbcommand: String,
    extraMessage: String?,
    status: String
) {
    open var contactMan: String = chests.first().finder

    open var chests: MutableList<ChChestData> = ArrayList(chests)

    @JvmField
    var bbcommand: String

    @JvmField
    var extraMessage: String?

    @Transient
    lateinit var color: Color

    @JvmField
    var lobbyId: Int = 0

    @JvmField
    var serverId: String
    lateinit var status: String
    var playersStillIn: MutableList<String>? = ArrayList()
    var closingTime: Instant = Instant.now().plus(360, ChronoUnit.MINUTES)
        protected set

    init {
        this.serverId = serverId
        this.bbcommand = bbcommand
        this.extraMessage = extraMessage
        setStatusNoOverride(status)
    }

    constructor(
        chest: MutableList<ChChestData>,
        serverId: String,
        bbcommand: String,
        extraMessage: String?,
        status: StatusConstants
    ) : this(
        chest,
        serverId,
        bbcommand,
        extraMessage,
        status.displayName
    ) {
        this.color = status.color
    }

    @Throws(SQLException::class)
    open fun setStatus(statusBase: StatusConstants) {
        setStatusNoOverride(statusBase)
    }

    fun setStatusNoOverride(statusBase: String) {
        this.status = statusBase
    }

    fun setStatusNoOverride(statusBase: StatusConstants) {
        this.status = statusBase.displayName
        color = statusBase.color

    }

    fun addChest(chest: ChChestData) {
        chests.add(chest)
    }

    fun transferToUser(newContactMan: String) {
        contactMan = newContactMan
    }

    @Throws(SQLException::class)
    open fun setLobbyMetaData(playersStillIn: MutableList<String>?, closingTime: Instant?) {
        this.playersStillIn = playersStillIn
        if (closingTime != null) {
            this.closingTime = closingTime
            onLobbyUpdate()
        }
    }

    open fun onLobbyUpdate() {
        // need to be overridden
    }

    override fun equals(other: Any?): Boolean {
        if (other !is ChestLobbyData) return false
        return other.lobbyId == lobbyId
    }

    protected open fun updateLobby(lobby: ChestLobbyData) {
        bbcommand = lobby.bbcommand
        extraMessage = lobby.extraMessage
        status = lobby.status
        contactMan = lobby.contactMan
        chests = ArrayList(lobby.chests)
        try {
            setLobbyMetaData(lobby.playersStillIn, lobby.closingTime)
        } catch (ignored: SQLException) {
        }
    }

    override fun hashCode(): Int {
        return lobbyId
    }
}
